Sources/_StringProcessing/Capture.swift

```diff
@@ -32,21 +32,42 @@ struct StoredCapture {
   var value: Any? = nil
 }
 
+// TODO: Where should this live? Inside TypeConstruction?
+func constructExistentialMatchComponent(
+  from input: Substring,
+  in range: Range<String.Index>?,
+  value: Any?,
+  optionalCount: Int
+) -> Any {
+  let someCount: Int
+  var underlying: Any
+  if let v = value {
+    underlying = v
+    someCount = optionalCount
+  } else if let r = range {
+    underlying = input[r]
+    someCount = optionalCount
+  } else {
+    // Ok since we Any-box every step up the ladder
+    underlying = Optional<Any>(nil) as Any
+    someCount = optionalCount - 1
+  }
+
+  for _ in 0..<someCount {
+    underlying = Optional(underlying) as Any
+  }
```

Comment:
The recursive type erasure is unfortunate and will make the final conversion (i.e. `as! Match`) slower.  This could've opened the existentials so that only the outer result is type-erased.
```swift
  for _ in 0..<someCount {
    func wrap<T>(_ x: T) {
      underlying = Optional(x) as Any
    }
    _openExistential(underlying, do: wrap)
  }
```

I've applied this in #182.